### ISSUE TITLE
only-arrow-functions rule should allow named functions

### ISSUE THREAD

**Issue by glen-84** [palantir/tslint#1449]
### Bug Report

- __TSLint version__: `3.14.0`
- __TypeScript version__: `1.8.10`
- __Running TSLint via__: n/a

#### TypeScript code being linted

```ts
function configureContainer(container: Container) {
    // ...
}
```

with `tslint.json` configuration:

```json
"only-arrow-functions": true
```

#### Actual behavior

`non-arrow functions are forbidden`

#### Expected behavior

Named functions should be allowed.

**Comment by JoshuaKGoldberg** [comment]
I can take a look at this within a day or two.

**Comment by JoshuaKGoldberg** [comment]
@glen-84 Does #1452 work for you? I went with adding an option for standalone declarations rather than all named functions because of cases such as:

```typescript
[1, 2, 3].filter(function whyIsThisNamed(num) { return num % 2 === 0; });
                          ^^^^^^^^^^^^^^
```

**Comment by glen-84** [comment]
LGTM =)

**Comment by JKillian** [comment]
Is there a benefit here to use a named function as opposed to `const functionName = () => { //... }`?

**Comment by glen-84** [comment]
Hmm. I'm not sure. :worried:

**Comment by JKillian** [comment]
For example, we'd want the rule to prevent the error case below:

```
class A {
	constructor(private b) { }
	
	public getMethod() {
		function foo() { console.log(this.b); }
		const bar = () => { console.log(this.b); }

		// return foo;    incorrect :(
		return bar;    // correct!
	}
}
```

I'm inclined not to add the extra complexity of an `allow-declarations` option, but if someone has a strong argument for it, I'm open to changing my mind

**Comment by danvk** [comment]
@JKillian The very-popular AirBnb style uses a setting like this:
https://github.com/airbnb/javascript#arrow-functions

**Comment by glen-84** [comment]
@JKillian @danvk is right, and it's also more consistent with the ESLint and JSCS rules, that focus on callbacks.

**Comment by JoshuaKGoldberg** [comment]
You can get some performance benefits to not having to create a new lambda to preserve scope.

Consider the case of a library that allows execution of callbacks with scope:

```typescript
class Executor {
    public setCallback(callback: Function, args?: any[]): void {
        this.callback = callback;
        this.args = args;
    }

    public setCallbackBound(callback: Function, scope?: any, args?: any[]): void {
        this.callback = callback;
        this.scope = scope;
        this.args = args;
    } 

    public execute(): void {
        this.callback.apply(this.scope, args);
    }
}
```

In using `setCallback`, which would necessitate an arrow lambda to preserve scope, you incur the runtime cost of creating a new scope.

```typescript
executor.setCallback(() => this.doStuff());
// ...
executor.execute();
```

In using `setCallbackBound` you just pass pointers around. It's a little more efficient.

```typescript
executor.setCallbackBound(this.doStuff, this);
// ...
executor.execute();
```

Because arrow lambdas capture scope you can't use them for places like this where scope is flexible.

**Comment by danvk** [comment]
@JoshuaKGoldberg you don't have to turn the rule on if it doesn't make sense for your application! In general, if you're using a library (e.g. jQuery) that sets `this` for its callbacks, then this rule is not for you.

**Comment by glen-84** [comment]
@danvk Joshua created the rule. =)

**Comment by JoshuaKGoldberg** [comment]
@danvk true! I see libraries like that as a reason why it could be useful to add flexibility to the rule.

**Comment by danvk** [comment]
Indeed! Naming the callback seems like a reasonable enough escape hatch from this rule.

**Comment by JKillian** [comment]
Okay, I'm good with this, full-steam ahead on the PR 🚢
